tokio/src/sync/semaphore.rs

```diff
@@ -123,6 +123,117 @@ use std::sync::Arc;
+/// # use std::collections::HashMap;
+/// # struct Database {
+/// #   map: HashMap<String, i32>,
+/// # }
+/// # impl Database {
+/// #    pub fn setup() -> Database {
+/// #        Database {
+/// #            map: HashMap::new(),
+/// #        }
+/// #    }
+/// #    pub fn insert(&mut self, key: &str, value: i32) {
+/// #        self.map.insert(key.to_string(), value);
+/// #    }
+/// #    pub fn update(&mut self, key: &str, value: i32) {
+/// #        self.map
+/// #            .entry(key.to_string())
+/// #            .and_modify(|origin| *origin = value);
+/// #    }
+/// #    pub fn delete(&mut self, key: &str) {
+/// #        self.map.remove(key);
+/// #    }
+/// #    pub fn get(&mut self, key: &str) -> i32 {
+/// #        *self.map.get(key).unwrap()
+/// #    }
+/// # }
+/// // Initialize a static semaphore with only one permit, which is used to
+/// // prevent test_insert and test_update from running in parallel.
+/// static PERMIT: Semaphore = Semaphore::const_new(1);
+///
+/// // Initialize the database that will be used by the subsequent tests.
+/// lazy_static! {
+///     static ref DB: Mutex<Database> = Mutex::new(Database::setup());
+/// }
+///
+/// async fn test_insert() {
+///
+///     // Acquire permit before proceeding. Since the semaphore has only one permit,
+///     // the test will wait if the permit is already acquired by other tests.
+///     let permit = PERMIT.acquire().await.unwrap();
+///
+///     // Do the actual test stuff with database
+///
+///     // Insert a key-value pair to database
+///     let (key, value) = ("name", 0);
+///     DB.lock().await.insert(key, value);
+///
+///     // Verify if the value has been inserted correctly.
+///     assert_eq!(DB.lock().await.get(key), value);
+///
+///     // Undo the insertation, make database status unchanged.
+///     DB.lock().await.delete(key);
+///
+///     // Drop permit, make possible waiting test to run.
+///     drop(permit);
+/// }
+///
+/// async fn test_update() {
+///
+///     // Acquire permit before proceeding. Since the semaphore has only one permit,
+///     // the test will wait if the permit is already acquired by other tests.
+///     let permit = PERMIT.acquire().await.unwrap();
+///
+///     // Do the same insert.
+///     let (key, value) = ("name", 0);
+///     DB.lock().await.insert(key, value);
+///
+///     // Update the existing value with a new one.
+///     let new_value = 1;
+///     DB.lock().await.update(key, new_value);
+///
+///     // Verify if the value has been updated correctly.
+///     assert_eq!(DB.lock().await.get(key), new_value);
+///
+///     // Undo any modificattion.
+///     DB.lock().await.delete(key);
+///
+///     // Drop permit, make possible waiting test to run.
+///     drop(permit);
+/// }
+///
+/// async fn test_others() {
+///     // this test can run in parallel with test_insert and test_update.
+///     // They have nothing to do with PERMIT.
+/// }
+/// # #[test]
+/// # fn fake_test() {}
```

Comment:
Please put this on the actual tests. The goal is to make `#[test]` show up as annotations on `test_insert` and `test_update`.

Reply:
Put the `#[test]` at `test_insert`:
```
/// #[test]
/// async fn test_insert() 
```

got the error:
```
error[E0425]: cannot find function `test_insert` in this scope
  --> tokio/src/sync/semaphore.rs:231:1
   |
94 | test_insert().await;
   | ^^^^^^^^^^^ not found in this scope
```